pkg/arvo/lib/pull-hook.hoon

```diff
@@ -295,72 +312,310 @@
+      =:  ship  s
+          rid   r
+          status  [%active ~]
+        ==
+      (tr-watch /)
+    ::
+    ++  tr-remove
+      tr-leave:tr-cleanup
+    ::
+    ++  tr-hook-act
+      |=  =action
+      ^+  tr-core
+      ?-  -.action
+        %add  (tr-add +.action)
+        %remove  tr-remove:(tr-abed resource.action)
+      ==
+    ::
+    ++  tr-cleanup
+      =.  gone  %.y
+      (tr-emit give-update)
+    ::
+    ++  tr-failed-kick
+      tr-core(status [%failed-kick ~])
+    ::
+    ++  tr-suspend-pub-ver
+      |=  ver=@ud
+      =.  status  [%pub-ver ver]
+      tr-leave:tr-watch-ver
+    ::
+    ::  
+    ++  tr-suspend-sub-ver
+      |=  ver=@ud
+      tr-core(status [%sub-ver ver])
+    ::
+    ++  tr-on-load
+      ?+  -.status  tr-core
+        %failed-kick  tr-restart
+        %active       tr-rewatch
+        ::
+          %sub-ver
+        ?.  (supported:ver (append-version:ver ver.status))
+          tr-core
+        tr-restart
+      ==
+    ::
+    ++  tr-restart
+      =.  status  [%active ~]
+      tr-kick
+    ::
+    ++  tr-rewatch
+      tr-kick:tr-leave
+    ::
+    ::
+    ::  +|  %subscription: subscription cards
+    ::
+    ::
+    ++  tr-ver-wire
+      (make-wire /version)
+    ::
+    ++  tr-watch-ver
+      (tr-emit (watch-version ship))
+    ::
+    ++  tr-leave-ver
+      (tr-emit (~(leave pass tr-ver-wire) tr-sub-dock))
+    ++  tr-sub-wire
+      (make-wire pull+resource+(en-path:resource rid))
+    ++  tr-unver-sub-wire
+      (make-wire pull+unver-resource+(en-path:resource rid))
+    ::
+    ++  tr-sub-dock
+      ^-  dock
+      [ship push-hook-name.config]
+    ::
+    ++  tr-check-sub
+      ?:  (~(has by wex.bowl) [tr-sub-wire tr-sub-dock])
+        tr-core
+      tr-kick
+    ::
+    ++  tr-watch-unver
+      |=  pax=path
+      =/  =path
+        :-  %resource
+        (weld (en-path:resource rid) pax)
+      (tr-emit (~(watch pass tr-unver-sub-wire) tr-sub-dock path))
+    ::
+    ++  tr-watch
+      |=  pax=path
+      ^+  tr-core
+      =/  =path
+        :+  %resource  %ver
+        %+  weld
+          (snoc (en-path:resource rid) (scot %ud version.config))
+        pax
+      (tr-emit (~(watch pass tr-sub-wire) tr-sub-dock path))
+    ::
+    ++  tr-leave
+      (tr-emit (~(leave pass tr-sub-wire) tr-sub-dock))
+    --
+  ::
+  ++  take-version
+    |=  [who=ship =sign:agent:gall]
+    ^-  [[(list card) _pull-hook] _state]
+    ?+  -.sign  !!
+        %watch-ack
+      ?~  p.sign  [~^pull-hook state]
+      =/  =tank  leaf+"subscribe failed from {<dap.bowl>} on wire {<wire>}"
+      %-  (slog tank u.p.sign)
+      [~^pull-hook state]
+      ::
+        %kick
+      :_  state
+      [(watch-version who)^~ pull-hook]
+      ::
+        %fact
+      ?.  =(%version p.cage.sign)
+        [~^pull-hook state]
+      =+  !<(version=@ud q.cage.sign)
+      =/  tracks=(list [rid=resource =track])
+        ~(tap by tracking)
+      =|  cards=(list card)
+      =|  leave=_&
```

Comment:
Doesn't `?` already bunt to `%.y`???

Reply:
Yes, but I find this behaviour confusing, and prefer to make it explicit